### ISSUE TITLE
Ensure identical deterministic build results and update release process

### ISSUE THREAD

**Issue by ageis** [zcash/zcash#1549]
#1521 and #1541 added deterministic builds with Gitian.

@str4d and @bitcartel compared manifests, and there was a diff in `cache_package_manifest` (the depends). 

This ticket is open to ensure that we run the Gitian builds using the tagged rc1 release, generate the same binaries, push our signatures to [gitian.sigs](https://github.com/zcash/gitian.sigs) and then we update the release process document for the Gitian builds and binaries for rc2.

**Comment by str4d** [comment]
@ageis and I both ran Gitian builds for the tagged rc1 release, and generated identical manifests:

https://github.com/zcash/gitian.sigs/blob/master/v1.0.0-rc1/kevin/zcash-1.0.0-rc1-build.assert
https://github.com/zcash/gitian.sigs/blob/master/v1.0.0-rc1/jack%40z.cash/zcash-1.0.0-rc1-build.assert

**Comment by ageis** [comment]
👍 I know I've been harping on this, but if we're serious about this, I'd say we need at least three devs publishing the build.assert and sig per release.

**Comment by radix42** [comment]
Do you guys want someone who isn't employed by zcashco to do and sign gitian builds for your releases, too? I'm assuming that the answer to that would be a "yes"....I could spin up a system on AWS that's big enough to do so in a reasonable amount of time that I only fire up for tagged releases.

**Comment by daira** [comment]
@radix42 Yes please!

**Comment by radix42** [comment]
I haven't run vagrant or gitian before, so it might take a while for me to get it all setup

**Comment by ageis** [comment]
@radix42 absolutely. We've automated a particular build environment setup in [our code](https://github.com/zcash/zcash-gitian), so it's not a lot of work for you other than installing prerequisites and running a couple commands -- less than 30 min until you're running the Gitian builds, even. If you want to use other virtualization platforms with gitian-builder or end up running into trouble with VirtualBox on AWS, you're _kinda_ on your own, however contact me if you have specific questions. It's good if we have more diverse build environments (for example, using KVM or Vbox instead of LXC for the actual build container), but at a minimum they should be Debian jessie based.

**Comment by radix42** [comment]
@ageis I'm not feeling up to rebasing my Windows port onto rc1 at the moment, so I'll work on getting this setup in my AWS account. There are a few spots in https://github.com/zcash/zcash-gitian that are ambiguous....should I add any questions I have about any roadblocks I hit to this ticket or a new/different one?

**Comment by ageis** [comment]
@radix42 you're welcome to also message me on the community Slack if you like. We tried to make the README clear to follow. :)

**Comment by ageis** [comment]
@radix42 you're welcome to also message me on the community Slack if you like. We tried to make the README clear to follow. :)

**Comment by radix42** [comment]
I get different results when I do a gitian build of rc1 (6BE5B036 is my gpg key ID):

```
vagrant@zcash-build:~/gitian.sigs/v1.0.0-rc1$ diff 6BE5B036/zcash-1.0.0-rc1-build.assert jack@z.cash/zcash-1.0.0-rc1-build.assert
4,5c4,5
<     91d381342f8376dedc727015d16cdb4ec232a2cf4c3d97a9f1acae919127a7f2  zcash-1.0.0-rc1-linux64-debug.tar.gz
<     42e1041fa916cca4ab559bb37abf6e6324c62952bffc85df64be053e4eaca730  zcash-1.0.0-rc1-linux64.tar.gz
---
>     267bed0369f8add2b0962cac944390db7bba5d78b588898558ff21dc963eb352  zcash-1.0.0-rc1-linux64-debug.tar.gz
>     60c4a368d0010924e3c3f168d1328674603598629dd3822ffa06d32703f7e812  zcash-1.0.0-rc1-linux64.tar.gz
341,342c341,342
<     a0961ad8c419975d3094843f8fb37511740fee78417125afef0d3ad4eff38ce7  x86_64-unknown-linux-gnu/libsnark/libsnark-0.1-bd7eaedfe8d.tar.gz
<     9aecb774d8c36840bdd70a8b50a9b2ac448f4ad9bbd7181a6638a934ec44dcc2  x86_64-unknown-linux-gnu/libsnark/libsnark-0.1-bd7eaedfe8d.tar.gz.hash
---
>     5db162459687cd350eca3e6927c65e9ab34980a860b9f3dbaf19cbf09ef3b304  x86_64-unknown-linux-gnu/libsnark/libsnark-0.1-bd7eaedfe8d.tar.gz
>     a3dcb0ba421c678e364148aed86ca9d7354bab31f5bacf15aa9c75cc9d308992  x86_64-unknown-linux-gnu/libsnark/libsnark-0.1-bd7eaedfe8d.tar.gz.hash
```

**Comment by ageis** [comment]
@radix42 Can you please find and scp zcash-1.0.0-rc1-linux64.tar.gz + libsnark-0.1-bd7eaedfe8d.tar.gz files out of the VM and try to attach it here or on Slack.

**Comment by radix42** [comment]
@ageis Here they are. Also,  my gitian.yml is below, can you see what I set wrong that makes my gpg ID show up rather than my email for my key, as you guys have on your sigs?
[libsnark-0.1-bd7eaedfe8d.tar.gz](https://github.com/zcash/zcash/files/539722/libsnark-0.1-bd7eaedfe8d.tar.gz)
[zcash-1.0.0-rc1-linux64.tar.gz](https://github.com/zcash/zcash/files/539723/zcash-1.0.0-rc1-linux64.tar.gz)


gitian.yml (it won't let you attach a yml file):
```
---
- name: Apply the Zcash Gitian builder role.
  become: yes
  hosts: localhost:zcash-build
  vars:
    zcash_git_repo_url: https://github.com/zcash/zcash
    zcash_version: v1.0.0-rc1
    gpg_key_name: '6BE5B036'
    git_name: 'David Mercer'
    git_email: 'radix42@gmail.com'
    gpg_key_id: '6BE5B036' # optional
    ssh_key_name: 'id_rsa' # optional
  roles:
    - role: common
      tags: common
    - role: gitian
      tags: gitian
```

**Comment by str4d** [comment]
Here's the diff of my and @radix42's libsnark builds:

[libsnark-rc1-str4d-vs-radix42.diff.txt](https://github.com/zcash/zcash/files/539797/libsnark-rc1-str4d-vs-radix42.diff.txt)

**Comment by radix42** [comment]
@ageis @str4d Here's my build.log, perhaps a diff of that and one of yours will reveal what's up with libsnark in my gitian build of rc1
[radix42-gitian-rc1-build.log.zip](https://github.com/zcash/zcash/files/540008/radix42-gitian-rc1-build.log.zip)

**Comment by str4d** [comment]
Here's mine:

[str4d-gitian-rc1-build.log.zip](https://github.com/zcash/zcash/files/540062/str4d-gitian-rc1-build.log.zip)

**Comment by radix42** [comment]
I think I found the difference:

```
$ diff str4d/build.log radix42/build.log

1797c1797
< checking build system type... sandybridge-pc-linux-gnu
---
> checking build system type... westmere-pc-linux-gnu
```

**Comment by radix42** [comment]
@ageis  @str4d  I think if we pass in --build=x86_64-unknown-linux-gnu to ./configure in gitian, libsnark should always build the same....the host is being passed in via the HOST environment variable, but it is figuring out the most specific processor it can generate code for, and this seems to make a difference when building libsnark. We'll want to do this, as per https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html the libsnark binary being built right now by gitian may or may not even RUN on a processor with a different specific instruction set (its optimizing too much!)

**Comment by radix42** [comment]
I myself would consider this a blocker until its fixed

**Comment by radix42** [comment]
I'm making a branch in my fork, I'm pretty sure we just need to pass in --host and --build to ./configure in build.sh explicitly, and that should make gitian builds the same regardless of the underlying processor architecture (provided of course that it is some flavor of x86_64).

**Comment by radix42** [comment]
Testing a further (one character!) patch to depends/Makefile so it respects both HOST and BUILD options passed in by build.sh. As soon as I have results from this gitian build, I'll submit a patch to upstream (you guys), assuming this works. It turns out that it is libgmp that is detecting the underlying processor instruction set at a finer level of detail than typical for gcc builds, and THAT is causing the differences in libsnark.

**Comment by radix42** [comment]
My pull request 1577 should remove the remaining indeterminism from v1.0.0-rc1 builds in libgmp and libsnark

**Comment by radix42** [comment]
I just verified that PR 1577 does indeed remove processor/chipset architecture from the build....of course you'll want to run it through the full test suite.

**Comment by daira** [comment]
Assuming fixed by #1557.
